dev-conventions.md

```diff
@@ -0,0 +1,227 @@
+   Before you open a PR, make sure to clean up your commit history and create the commits that you think best divide up the total work as outlined above (use `git rebase` and `git commit --amend`).
+
+There are three ways of 'merging' pull requests on GitHub:
+
+- **Squash and merge**: take all commits, squash them into a single one and put it on top of the base branch.
+  - Choose this by default for pull requests that address a single issue and are well represented by a single commit.
+  - Make sure to clean the commit message (title & body)
+- **Rebase and merge**: take all commits and 'recreate' them on top of the base branch. All commits will be recreated with new hashes.
+  - Choose this for pull requests that require more than a single commit.
+  - Examples: PRs that contain multiple commits with individually significant changes; PRs that have commits from different authors (squashing commits would remove attribution)
+- **Merge with merge commit**: put all commits as they are on the base branch, with a merge commit on top
+  - Choose for collaborative PRs with many commits.
+     Here, the merge commit provides actual benefits.
+
+For the latter two, it is recommend that the merger do (if it is their PR) or at ask the contributor to perform an [interactive rebase]( https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history). At this point you can squash consecutive commits together (using "fixup") and also rewrite commit messages (using "reword") to create a compliant history.
+
+(dev/commits)=
+
+## Commit Messages
+
+A commit:
+
+- should ideally address one issue
+- should be a self-contained change to the code base
+- must not lump together unrelated changes
+
+Repositories should follow the following convention (where possible):
+
+```markdown
+<EMOJI> <KEYWORD>: Summarize changes in 72 characters or less
+
+More detailed explanatory text, if necessary.
+The blank line separating the summary from the body is
+critical (unless you omit the body entirely); various tools like `log`,
+`shortlog` and `rebase` can get confused if you run the two together.
+
+Explain the problem that this commit is solving. Focus on why you are
+making this change as opposed to how (the code explains that).  Are
+there side effects or other unintuitive consequences of this change?
+Here's the place to explain them to someone else reading your message in
+the future. Make sure to include some necessary context for difficult
+concepts that might change in the future.
+
+There is no need to mention you also added unit tests when adding a new feature. The code diff already makes this clear.
+```
+
+Keywords/emojis are adapted from [Emoji-Log](https://github.com/ahmadawais/Emoji-Log) and [gitmoji](https://github.com/carloscuesta/gitmoji) and should be one of the following (brackets contain [GitHub emoji markup](https://gist.github.com/rxaviers/7360908)):
+
+- `✨  NEW:` (`:sparkles:`) — to introduce a new feature(s).
```

Comment:
just to clarify we should be pasting in emojis *directly* rather than putting `:sparkles:` in the commit message itself, right? So we would type sparkles into a comment window, preview the comment, copy/paste the emoji, and then insert it into the message? (or, directly paste using an emoji keyboard or something)

Reply:
yep directly the emoji (GitHub will auto-replace IRC, if you type it in the PR commit window)